internal/column-helpers/columnhelpers.go

```diff
@@ -0,0 +1,307 @@
+
+	// SetErrorCallback sets the function to use whenever errors occur in this library or are received in-band
+	SetErrorCallback(errorCallback ErrorCallback)
+
+	// EnableSnapshots initializes the snapshot collector, which is able to aggregate snapshots from several sources
+	// and can return (optionally cached) results on demand; used for top gadgets
+	EnableSnapshots(ctx context.Context, t time.Duration, ttl int)
+}
+
+type ColumnHelpers[T any] struct {
+	columns            *columns.Columns[T]
+	sortBy             []string
+	filters            []string
+	filterSpec         *filter.FilterSpec[T] // TODO: filter collection(!)
+	eventCallback      func(*T)
+	eventCallbackArray func([]*T)
+	errorCallback      ErrorCallback
+	snapshotCombiner   *snapshotcombiner.SnapshotCombiner[T]
+	enableSnapshots    bool
+	columnFilters      []columns.ColumnFilter
+}
+
+func NewColumnHelpers[T any](columns *columns.Columns[T]) Columns {
+	ch := &ColumnHelpers[T]{
+		columns: columns,
+	}
+	return ch
+}
+
+func (ch *ColumnHelpers[T]) EnableSnapshots(ctx context.Context, interval time.Duration, ttl int) {
+	if ch.eventCallbackArray == nil {
+		panic("EnableSnapshots needs EventCallbackArray set")
+	}
+	ch.snapshotCombiner = snapshotcombiner.NewSnapshotCombiner[T](ttl)
+	ch.enableSnapshots = true
+	go func() {
+		ticker := time.NewTicker(interval)
+		for {
+			select {
+			case <-ticker.C:
+				out, _ := ch.snapshotCombiner.GetSnapshots()
+				ch.eventCallbackArray(out)
+			case <-ctx.Done():
+			}
+		}
+	}()
+}
+
+func (ch *ColumnHelpers[T]) SetColumnFilters(filters ...columns.ColumnFilter) {
+	ch.columnFilters = filters
+}
+
+func (ch *ColumnHelpers[T]) SetErrorCallback(errorCallback ErrorCallback) {
+	ch.errorCallback = errorCallback
+}
+
+func (ch *ColumnHelpers[T]) SetEventCallback(eventCallback any) {
+	// Check, whether the caller wants a generic callback function (e.g. to print JSON)
+	gcb, ok := eventCallback.(func(any))
+	if ok {
+		ch.eventCallback = func(ev *T) {
+			gcb(ev)
+		}
+		return
+	}
+
+	// If it's non-generic, we expect it to know our type T
+	cb, ok := eventCallback.(func(*T))
+	if !ok {
+		panic("cannot use event callback for columnhelper")
+	}
+	ch.eventCallback = cb
+}
+
+func (ch *ColumnHelpers[T]) SetEventCallbackArray(eventCallback any) {
+	// Check, whether the caller wants a generic callback function (e.g. to print JSON)
+	gcb, ok := eventCallback.(func(any))
+	if ok {
+		ch.eventCallbackArray = func(ev []*T) {
+			gcb(ev)
+		}
+		return
+	}
+
+	// If it's non-generic, we expect it to know our type T
+	cb, ok := eventCallback.(func([]*T))
+	if !ok {
+		panic("cannot use event callback for columnhelper")
+	}
+	ch.eventCallbackArray = cb
+}
+
+func (ch *ColumnHelpers[T]) eventHandler(enrichers ...func(any)) func(*T) {
+	return func(ev *T) {
+		if enrichers != nil {
+			for _, enricher := range enrichers {
+				enricher(ev)
+			}
+		}
+		if ch.filterSpec != nil && !ch.filterSpec.Match(ev) {
+			return
+		}
```

Comment:
This is tough On one hand one could avoid calling the enrichers by filtering beforehand, but on the other hand you might want to filter on data that has been added by enriching... So I guess the filtering should always come last to be able to catch everything.
Filtering using eBPF is of course unaffected and handled before. We can probably improve things later on by defining what exactly could be filtered before enriching (because data won't change) and what not.